src/dma/mod.rs

```diff
@@ -0,0 +1,1171 @@
+        /// Set the transfer_error_interrupt.
+        #[inline]
+        pub fn transfer_error_interrupt(mut self, transfer_error_interrupt: bool) -> Self {
+            self.transfer_error_interrupt = transfer_error_interrupt;
+            self
+        }
+        /// Set the direct_mode_error_interrupt.
+        #[inline]
+        pub fn direct_mode_error_interrupt(mut self, direct_mode_error_interrupt: bool) -> Self {
+            self.direct_mode_error_interrupt = direct_mode_error_interrupt;
+            self
+        }
+        /// Set the fifo_error_interrupt.
+        #[inline]
+        pub fn fifo_error_interrupt(mut self, fifo_error_interrupt: bool) -> Self {
+            self.fifo_error_interrupt = fifo_error_interrupt;
+            self
+        }
+        /// Set the double_buffer.
+        #[inline]
+        pub fn double_buffer(mut self, double_buffer: bool) -> Self {
+            self.double_buffer = double_buffer;
+            self
+        }
+        /// Set the fifo_threshold.
+        #[inline]
+        pub fn fifo_threshold(mut self, fifo_threshold: FifoThreshold) -> Self {
+            self.fifo_threshold = fifo_threshold;
+            self
+        }
+        /// Set the fifo_enable.
+        #[inline]
+        pub fn fifo_enable(mut self, fifo_enable: bool) -> Self {
+            self.fifo_enable = fifo_enable;
+            self
+        }
+        /// Set the memory_burst.
+        #[inline]
+        pub fn memory_burst(mut self, memory_burst: BurstMode) -> Self {
+            self.memory_burst = memory_burst;
+            self
+        }
+        /// Set the peripheral_burst.
+        #[inline]
+        pub fn peripheral_burst(mut self, peripheral_burst: BurstMode) -> Self {
+            self.peripheral_burst = peripheral_burst;
+            self
+        }
+    }
+}
+
+/// DMA Transfer.
+pub struct Transfer<STREAM, CHANNEL, PERIPHERAL, DIRECTION, BUF>
+where
+    STREAM: Stream,
+    PERIPHERAL: PeriAddress,
+    BUF: WriteBuffer<Word = <PERIPHERAL as PeriAddress>::MemSize> + 'static,
+{
+    stream: STREAM,
+    _channel: PhantomData<CHANNEL>,
+    peripheral: PERIPHERAL,
+    _direction: PhantomData<DIRECTION>,
+    buf: Option<BUF>,
+    double_buf: Option<BUF>,
+    // Used when double buffering
+    transfer_length: u16,
+}
+
+impl<STREAM, CHANNEL, PERIPHERAL, DIR, BUF> Transfer<STREAM, CHANNEL, PERIPHERAL, DIR, BUF>
+where
+    STREAM: Stream,
+    CHANNEL: Channel,
+    DIR: Direction,
+    PERIPHERAL: PeriAddress,
+    BUF: WriteBuffer<Word = <PERIPHERAL as PeriAddress>::MemSize> + 'static,
+    (STREAM, CHANNEL, PERIPHERAL, DIR): Sealed,
+{
+    /// Applies all fields in DmaConfig.
+    fn apply_config(&mut self, config: config::DmaConfig) {
+        let msize = mem::align_of::<<PERIPHERAL as PeriAddress>::MemSize>() / 2;
+
+        self.stream.clear_interrupts();
+        self.stream.set_priority(config.priority);
+        // NOTE(unsafe) These values are correct because of the invariants of PeriAddress
```

Comment:
Only the HAL can implement `PeriAddress` because of the `Sealed` bound, but maybe we should mark as `unsafe` to be more concise and maybe help other people extending this API.

Reply:
Can we really use "it's only used internally" as an excuse for not marking code that's potentially unsafe? I can't remember dealing with such cases before, but I always thought of safety contracts and accessibility as orthogonal concerns.

HAL developers from the future can make mistakes too, so adding that as an additional documentation is helpful I think. Plus, as you said, it's more future-proof that way.